Add local device state storage (also getEffect methods)

tildearrow opened the pull request:
#852. This means:
- When coming back from a Windows machine (where Synapse is installed) your devices will be restored to the last state (effect and settings like DPI).
- Frontends may use the get\*Effect methods (provided in the pylib as well) to properly display the current effect. Currently:
  - Polychromatic desyncs when changing the effect with a different frontend
  - razerCommander doesn't remember anything.
  - RazerGenie displays the wrong color and effect upon opening
- The getDPI, getPollRate and getBrightness methods (and possible any other getters) will no longer sporadically return 0.
- This allows for fixing the 3 issues mentioned above, and also fixes [this](https://github.com/polychromatic/polychromatic/issues/186), [this](https://github.com/polychromatic/polychromatic/issues/148), [this](https://github.com/polychromatic/polychromatic/issues/60), [this](https://github.com/polychromatic/polychromatic/issues/14), [this](https://github.com/polychromatic/polychromatic/issues/189), [this](https://github.com/openrazer/openrazer/issues/693#issuecomment-437977247), [this](https://github.com/openrazer/openrazer/issues/747#issuecomment-513495112), [this](https://github.com/openrazer/openrazer/issues/540#issuecomment-514605828), [this](https://github.com/GabMus/razerCommander/issues/49), [this](https://github.com/z3ntu/RazerGenie/issues/46), [this](https://github.com/z3ntu/RazerGenie/issues/30), [this](https://github.com/polychromatic/polychromatic/issues/216) and [this](https://github.com/openrazer/openrazer/issues/663), and some other issues I forgot to mention.

Possible problems:
- Resets your devices to Spectrum on first startup
- Not tested on BlackWidow Ultimate 2013, DeathAdder Chroma, Old Kraken and any device with hardware Starlight effect.
- The Old Kraken and DeathAdder Chroma **do** have effect retrieval methods (and even the Old Kraken has a method with the exact same name as one this PR provides (getCurrentEffect)), but I hope you can let me remove these as they'll become deprecated with this PR.

Addendum: Yes, it's best to do this in the daemon rather than in the frontends as:
- It works cross-frontend
- Works without any frontend at all as well

tildearrow committed:
initial implementation of fx storage/getEffect

z3ntu commented on daemon/openrazer_daemon/dbus_services/dbus_methods/chroma_keyboard.py:
```diff
@@ -402,10 +433,14 @@ def set_breath_triple_effect(self, red1, green1, blue1, red2, green2, blue2, red
+
+    # remember effect
+    self.current_effect = 'breathTriple'
+    self.current_effect_colors = int(red1), int(green1), int(blue1), int(red2), int(green2), int(blue2), int(red3), int(green3), int(blue3)
```
Shoudln't it be `self.current_effect_colors[0:9] = ` ?

tildearrow replied:
I thought the = operator assigned? The size of the current_effect_colors list is implicit, 9.

z3ntu commented on daemon/openrazer_daemon/dbus_services/dbus_methods/deathadder_chroma.py:
```diff
@@ -81,6 +81,8 @@ def get_logo_effect(self):
     :rtype: bool
     """
+    # TODO: remove. we have a new solution that works across all
+    # devices and is more reliable (getCurrentLogoEffect)
```
Could you grep through the applications listed in the readme to see if this method gets used?

tildearrow replied:
Polychromatic: Not used
razerCommander: Not used
Chroma Feedback: Not used
RazerGenie: Method exposed in libopenrazer, otherwise not used

Bonus! KeyboardVisualizer: Not used

tildearrow replied:
I will have to tweak the Kraken device suspend code in order to remove the Kraken-specific getCurrentEffect method...

tildearrow replied:
Done

z3ntu commented on daemon/openrazer_daemon/dbus_services/dbus_methods/deathadder_chroma.py:
```diff
@@ -377,7 +400,7 @@ def set_scroll_static(self, red, green, blue):
 def set_scroll_blinking(self, red, green, blue):
     """
-    Set the device to pulsate
+    Set the device to BLINKING
```
Make it lowercase as the rest of the comments

tildearrow replied:
Done.

z3ntu commented on daemon/openrazer_daemon/dbus_services/dbus_methods/deathadder_chroma.py:
```diff
@@ -435,22 +466,16 @@ def set_scroll_pulsate(self, red, green, blue):
-
-    :param blue: Blue component
-    :type blue: int
+    Set the device to copypaste
```
I saw that :wink:

tildearrow replied:
Done. Haha

z3ntu commented on daemon/openrazer_daemon/hardware/device_base.py:
```diff
@@ -104,13 +134,371 @@ def __init__(self, device_path, device_number, config, testing=False, additional
+            ('razer.device.misc', 'getCurrentLogoEffect', self.get_current_logo_effect, None, 's'),
+            ('razer.device.misc', 'getCurrentLogoEffectColors', self.get_current_logo_effect_colors, None, 'ay'),
+            ('razer.device.misc', 'getCurrentLogoEffectSpeed', self.get_current_logo_effect_speed, None, 'i'),
+        }
```
Shouldn't the other led locations also have the wave direction method?

tildearrow replied:
Done.

z3ntu commented on daemon/openrazer_daemon/hardware/device_base.py:
```diff
@@ -104,13 +134,371 @@ def __init__(self, device_path, device_number, config, testing=False, additional
             self.logger.debug("Adding {}.{} method to DBus".format(m[0], m[1]))
             self.add_dbus_method(m[0], m[1], m[2], in_signature=m[3], out_signature=m[4])
 
+        if 'set_static_effect' in self.METHODS:
```
This could break on some older devices that just have on/off and no rgb (see Abyssus 1800 for example)

tildearrow replied:
This effect storage pull request only supports Chroma devices and doesn't work for devices with the set_logo_active method.
(However, I can add support for it although it'll require testing since I don't own any devices from that era)

tildearrow replied:
Done with last set of commits that implement storage for LED status.

z3ntu commented on daemon/openrazer_daemon/hardware/device_base.py:
```diff
@@ -104,13 +134,371 @@ def __init__(self, device_path, device_number, config, testing=False, additional
 
+        # load last effect
+        if self.has_normal_effects:
+            if self.config.has_section(self._serial):
```
Some old devices use an temporary generated serial (see driver source), that should be maybe changed to always start with e.g. ABC000000 and then count up per devices plugged in so we can have some kind of certainty that the serial will exist again

tildearrow replied:
Done. Serial-less devices will be stored by name instead.

z3ntu commented on daemon/openrazer_daemon/hardware/device_base.py:
```diff
@@ -104,13 +134,371 @@ def __init__(self, device_path, device_number, config, testing=False, additional
+                elif effect_func_name == 'setStarlightSingle':
+                    effect_func(self.current_effect_speed, self.current_effect_colors[0], self.current_effect_colors[1], self.current_effect_colors[2])
+                elif effect_func_name == 'setStarlightDual':
+                    effect_func(self.current_effect_speed, self.current_effect_colors[0], self.current_effect_colors[1], self.current_effect_colors[2], self.current_effect_colors[3], self.current_effect_colors[4], self.current_effect_colors[5])
```
Would it be possible to make a more automatic system here where it detects the parameters of the function instead of hardcoding these?

tildearrow replied:
It may be possible, but I don't have that many Python skills...

tildearrow replied:
Done. However, Starlight will remain hard-coded because it has different arguments/argument order (reactive is RGBS, starlight single is SRGB) (starlight random is S, wave is W)

z3ntu commented on daemon/openrazer_daemon/hardware/device_base.py:
```diff
@@ -104,13 +134,371 @@ def __init__(self, device_path, device_number, config, testing=False, additional
+                    effect_func(self.current_effect_speed, self.current_effect_colors[0], self.current_effect_colors[1], self.current_effect_colors[2], self.current_effect_colors[3], self.current_effect_colors[4], self.current_effect_colors[5])
+
+        # load last logo effect
+        if self.has_logo_effects:
```
So much copy pasta :( It would be nice if you could try to unify those blocks into one that goes through 5 times with the different zones

tildearrow replied:
I never do nice things!
I will try using a map for the effect storage instead of independent variables to do so.

tildearrow replied:
Done through refactor

z3ntu commented on daemon/openrazer_daemon/hardware/keyboards.py:
```diff
@@ -55,6 +55,18 @@ def __init__(self, *args, **kwargs):
+                if effect_func_name == 'setRipple':
+                    effect_func(self.current_effect_colors[0], self.current_effect_colors[1], self.current_effect_colors[2], 0.025)
+                elif effect_func_name == 'setRippleRandomColour':
+                    effect_func(0.025)
```
I think we have some constants defined somewhere for the 0.025 for example

tildearrow replied:
Thanks for pointing out. I have fixed it now.
However, I do have to tell you that the constant is set to 0.04, which means 25 FPS, making the effect stutter slightly. Is it possible for me to raise the speed to 0.025? (40 FPS)

z3ntu reviewed (changes requested):
See comments, looks pretty good otherwise. I'll try to implement support for that in RazerGenie and I'll see how it goes. I might wait for Luke to try to implement that in polychromatic as well to make sure the APIs are good

tildearrow committed:
apply z3ntu-requested changes.

tildearrow committed:
use default values for ripple initialization

tildearrow committed:
prepare to refactor

tildearrow committed:
refactor, part 2

this will not work yet

tildearrow committed:
refactor, part 3

this is the last commit that refactors the code for less copy-paste
and better readability.

however, i'm not done yet...

tildearrow committed:
remove device-specific hardware effect retrieval methods

tildearrow committed:
CI

tildearrow committed:
really finish CI

tildearrow committed:
third time is the charm

tildearrow committed:
increase ripple default FPS

tildearrow committed:
more safety checks

tildearrow committed:
oh come on

tildearrow committed:
add support for old device light status storage

tildearrow committed:
finish effect storage for old devices

tildearrow committed:
are you being serious?

tildearrow committed:
✅

tildearrow committed:
fix effect storage for devices without serial

tildearrow committed:
automatic argument detection

tildearrow committed:
prepare for brightness storage

tildearrow committed:
finish brightness storage

no more random 0% brightness as we don't rely on the hardware
anymore for this

tildearrow committed:
MWTGM (merge what they got merged)

tildearrow committed:
Add support for effect storage

z3ntu commented on daemon/openrazer_daemon/dbus_services/dbus_methods/deathadder_chroma.py:
```diff
@@ -73,23 +79,6 @@ def set_logo_active(self, active):
 
 
-@endpoint('razer.device.lighting.logo', 'getLogoEffect', out_sig='y')
-def get_logo_effect(self):
```
This function is implemented in libopenrazer in the RazerGenie master branch but it isn't used. I've also verified that polychromatic and razerCommander don't use the functions that you've removed: getLogoEffect getScrollEffect getBreathArgs getStaticArgs getCurrentEffect ; so it should be fine removing them

tildearrow replied:
What does this mean? How can I resolve this conversation?

tildearrow replied:
No answer, assuming comment.

z3ntu commented on daemon/openrazer_daemon/dbus_services/dbus_methods/deathadder_chroma.py:
```diff
@@ -283,23 +287,6 @@ def set_scroll_active(self, active):
 
 
-@endpoint('razer.device.lighting.scroll', 'getScrollEffect', out_sig='y')
-def get_scroll_effect(self):
```
`get_scroll_effect` is still present two times in `daemon/openrazer_daemon/hardware/mouse.py`

tildearrow replied:
Fixed

z3ntu commented on daemon/openrazer_daemon/hardware/device_base.py:
```diff
@@ -63,6 +69,30 @@ def __init__(self, device_path, device_number, config, testing=False, additional
+        elif self.USB_PID == 0x0024 or self.USB_PID == 0x0025:
+            self.storage_name = "Mamba2012"
+        else:
+            self.storage_name = self.serial
```
Good catch handling the devices without a static serial

tildearrow replied:
What does this mean? Is this wrong?

tildearrow replied:
No answer, assuming it is a comment.

z3ntu commented on daemon/openrazer_daemon/hardware/device_base.py:
```diff
@@ -104,13 +134,176 @@ def __init__(self, device_path, device_number, config, testing=False, additional
             self.logger.debug("Adding {}.{} method to DBus".format(m[0], m[1]))
             self.add_dbus_method(m[0], m[1], m[2], in_signature=m[3], out_signature=m[4])
 
+        if 'set_static_effect' in self.METHODS:
```
Please add a comment why this check has to be done separately from the rest

tildearrow replied:
Done

z3ntu commented on daemon/openrazer_daemon/hardware/device_base.py:
```diff
@@ -104,13 +134,176 @@ def __init__(self, device_path, device_number, config, testing=False, additional
+                    try:
+                        self.zone[i]["effect"] = self.config[self.storage_name][i + '_effect']
+                    except KeyError:
+                        self.zone[i]["effect"] = 'spectrum'
```
Aren't these values already the default values set above? Same comment applies to the other `except` parts below as well

tildearrow replied:
Yep, sorry about that. Fixed.

tildearrow replied:
However, I didn't remove the re-initialization code in the `except ValueError` because it handles incomplete color reading (instead of leaving us with a broken color set, it resets the colors).

z3ntu commented on daemon/openrazer_daemon/hardware/device_base.py:
```diff
@@ -104,13 +134,176 @@ def __init__(self, device_path, device_number, config, testing=False, additional
+                    if i == "backlight":
+                        effect_func_name = 'set' + self.zone[i]["effect"][0].upper() + self.zone[i]["effect"][1:]
+                    else:
+                        effect_func_name = 'set' + i[0].upper() + i[1:] + self.zone[i]["effect"][0].upper() + self.zone[i]["effect"][1:]
```
Please add some comments to the lines above describing what it's doing (or try to simplify the code)

~~Also couldn't you use e.g. 'setBacklightStatic' for the backlight zones instead of 'setStatic' (at least that's what I understand is happening)~~ nevermind I understand now

Also there's the method [`str.capitalize()`](https://docs.python.org/3.7/library/stdtypes.html#str.capitalize) which does the `i[0].upper() + i[1:]` which makes the code more understandable so please use that wherever you used this construct

tildearrow replied:
Done.

**No, sorry. I will not use .capitalize() because I have a good reason not to:**
.capitalize() is NOT the same as `i[0].upper() + i[1:]`. It is more like `i[0].upper() + i[1:].lower()`. This means, as an example `breathRandom` will NOT be converted to `BreathRandom` but rather `Breathrandom`, which is not going to work. And so far as I know, there is no function to just capitalize the first letter. So I have to do this. Sorry.

z3ntu replied:
You're right about `.capitalize()` , but then please make a helper function to do that so this pattern doesn't have to exist 10 times in the code

tildearrow replied:
Hello. I tried.
```
daemon/openrazer_daemon/hardware/device_base.py:258:47: E0602: Undefined variable 'capitalize_first_char' (undefined-variable)
```
Why am I getting this error? I clearly declared the capitalize_first_char function as static under RazerDevice.

tildearrow replied:
Ooooh, figured out why. Done.

z3ntu commented on daemon/openrazer_daemon/hardware/device_base.py:
```diff
@@ -104,13 +134,176 @@ def __init__(self, device_path, device_number, config, testing=False, additional
+                    else:
+                        effect_func_name = 'set' + i[0].upper() + i[1:] + self.zone[i]["effect"][0].upper() + self.zone[i]["effect"][1:]
+                else:
+                    self.zone[i]["effect"] = 'spectrum'
```
This is again the default already

tildearrow replied:
Done

z3ntu commented on daemon/openrazer_daemon/hardware/device_base.py:
```diff
@@ -104,13 +134,176 @@ def __init__(self, device_path, device_number, config, testing=False, additional
+                    if i == "backlight":
+                        effect_func_name = 'setSpectrum'
+                    else:
+                        effect_func_name = 'set' + i[0].upper() + i[1:] + 'Spectrum'
```
At least one assignment can be get rid of here as in `i = "backlight"` `effect_func_name` gets assigned twice to the same value

tildearrow replied:
OK

z3ntu commented on daemon/openrazer_daemon/hardware/device_base.py:
```diff
@@ -104,13 +134,176 @@ def __init__(self, device_path, device_number, config, testing=False, additional
+
+                if not effect_func == None:
+                    if self.get_num_arguments(effect_func) == 1:
+                        if self.zone[i]["effect"] == 'starlightRandom':
```
I'd rather compare the `effect_func_name` as it's more clear instead of going back to the `effect` value

tildearrow replied:
Not doing this one, because you didn't realize one thing: it'd require a lot more code since `effect_func_name` may change per zone (e.g. `setSpectrum` and `setLogoSpectrum`)

z3ntu commented on daemon/openrazer_daemon/hardware/device_base.py:
```diff
@@ -104,13 +134,176 @@ def __init__(self, device_path, device_number, config, testing=False, additional
+                    effect_func = getattr(self, effect_func_name, None)
+
+                if not effect_func == None:
+                    if self.get_num_arguments(effect_func) == 1:
```
This section deserves a few comments e.g. `There are 2 functions having 1 argument only, setStarlightRandom and setWave`

tildearrow replied:
Done

z3ntu commented on daemon/openrazer_daemon/hardware/device_base.py:
```diff
@@ -104,13 +134,176 @@ def __init__(self, device_path, device_number, config, testing=False, additional
+                    if self.get_num_arguments(effect_func) == 1:
+                        if self.zone[i]["effect"] == 'starlightRandom':
+                            effect_func(self.zone[i]["speed"])
+                        else:
```
I wouldn't use an `else` here because even if there are currently only two functions with one param, this could break in the future so do a comparison for wave and have the else branch used for an error message for the log

tildearrow replied:
Done

z3ntu commented on daemon/openrazer_daemon/hardware/device_base.py:
```diff
@@ -104,13 +134,176 @@ def __init__(self, device_path, device_number, config, testing=False, additional
+                        if self.zone[i]["effect"] == 'starlightSingle':
+                            effect_func(self.zone[i]["speed"], self.zone[i]["colors"][0], self.zone[i]["colors"][1], self.zone[i]["colors"][2])
+                        else:
+                            effect_func(self.zone[i]["colors"][0], self.zone[i]["colors"][1], self.zone[i]["colors"][2], self.zone[i]["speed"])
```
Same thing about the else branch applies here

tildearrow replied:
Done

z3ntu commented on daemon/openrazer_daemon/hardware/device_base.py:
```diff
@@ -104,13 +134,176 @@ def __init__(self, device_path, device_number, config, testing=False, additional
+                    elif self.get_num_arguments(effect_func) == 9:
+                        effect_func(self.zone[i]["colors"][0], self.zone[i]["colors"][1], self.zone[i]["colors"][2], self.zone[i]["colors"][3], self.zone[i]["colors"][4], self.zone[i]["colors"][5], self.zone[i]["colors"][6], self.zone[i]["colors"][7], self.zone[i]["colors"][8])
+                    else:
+                        effect_func()
```
`elif self.get_num_arguments(effect_func) == 0` and put an error message in the else branch

tildearrow replied:
Done

z3ntu commented on daemon/openrazer_daemon/hardware/device_base.py:
```diff
@@ -104,13 +134,176 @@ def __init__(self, device_path, device_number, config, testing=False, additional
+                    elif self.get_num_arguments(effect_func) == 7:
+                        effect_func(self.zone[i]["speed"], self.zone[i]["colors"][0], self.zone[i]["colors"][1], self.zone[i]["colors"][2], self.zone[i]["colors"][3], self.zone[i]["colors"][4], self.zone[i]["colors"][5])
+                    elif self.get_num_arguments(effect_func) == 9:
+                        effect_func(self.zone[i]["colors"][0], self.zone[i]["colors"][1], self.zone[i]["colors"][2], self.zone[i]["colors"][3], self.zone[i]["colors"][4], self.zone[i]["colors"][5], self.zone[i]["colors"][6], self.zone[i]["colors"][7], self.zone[i]["colors"][8])
```
It wouldn't hurt putting `self.zone[i]` into a local variable and probably also putting the colors into their own local variable would make the code shorter

tildearrow replied:
Done.

z3ntu commented on daemon/openrazer_daemon/hardware/keyboards.py:
```diff
@@ -55,6 +55,18 @@ def __init__(self, *args, **kwargs):
+            effect_func_name = 'set' + self.zone["backlight"]["effect"][0].upper() + self.zone["backlight"]["effect"][1:]
+            effect_func = getattr(self, effect_func_name, None)
+
+            if not effect_func == None:
```
`if effect_func is not None:`

tildearrow replied:
Done, including device_base

z3ntu commented on daemon/openrazer_daemon/misc/ripple_effect.py:
```diff
@@ -25,7 +25,7 @@ def __init__(self, parent, device_number):
 
         self._colour = (0, 255, 0)
-        self._refresh_rate = 0.040
+        self._refresh_rate = 0.025
```
This looks like an unrelated change, please put it into a separate PR or at least into its own commit

tildearrow replied:
Fixed. 25 FPS looks terrible

tildearrow replied:
GitHub please stop hiding outdated conversations! I assume they are fixed when in fact they are not

z3ntu reviewed (changes requested):
Thanks a lot for updating the PR and progressing on this topic! See comments

tildearrow committed:
remove unused instances of "get_scroll_effect"

tildearrow committed:
oh sorry

tildearrow committed:
add comment

tildearrow committed:
remove duplicate variable init

tildearrow committed:
comment-ize and provide ugly stuff rationale

tildearrow committed:
more

tildearrow committed:
compare EFN instead of E

tildearrow committed:
Revert "compare EFN instead of E"

This reverts commit 63e3d8b43295254bfc6860014f6e88e463428b99.

tildearrow committed:
.

tildearrow committed:
is not None

tildearrow committed:
Revert refresh rate

tildearrow committed:
smaller code

tildearrow committed:
CIIIIIIIIIIIIIIIIIIIIIII

tildearrow committed:
Merge branch 'master' of https://github.com/openrazer/openrazer and fix starlight argument order

tildearrow committed:
this is really tiring me

tildearrow committed:
asdjklf

tildearrow committed:
yes. I finally found out why

tildearrow committed:
move getEffect methods to respective places

also change the pylib API regarding getEffect.
much cleaner

tildearrow committed:
rebase on master.

tildearrow committed:
fix an important bug

tildearrow committed:
Merge branch 'master' of https://github.com/openrazer/openrazer

tildearrow committed:
fix data storage and hotplug thingy

tildearrow committed:
NOTE: this does not work yet

tildearrow committed:
fix storage if keys not present

z3ntu commented on daemon/openrazer_daemon/daemon.py:
```diff
@@ -253,16 +253,48 @@ def read_config(self, config_file):
+        self._config['Startup']['devices_off_on_screensaver'] = 'True' if self.get_off_on_screensaver() else 'False'
+
+        for device in self._razer_devices:
+            self._config[device.dbus.storage_name] = {
```
`= {}` (without the newline) please

tildearrow replied:
Done

z3ntu commented on pylib/openrazer/client/fx.py:
```diff
@@ -68,6 +68,46 @@ def __init__(self, serial: str, capabilities: dict, daemon_dbus=None, matrix_dim
+
+        :return: Effect name
+        :rtype: str
+        """
```
Please add example strings to the doc

tildearrow replied:
Please explain I am not familiar with Python
Furthermore neither of the other functions provide examples so why is this necessary now?

z3ntu replied:
Sorry for not replying earlier, I meant `:return: Effect name (e.g. "spectrum")` for example, just so the user has an idea what the return value looks like without running the call

tildearrow replied:
Done

z3ntu commented on pylib/openrazer/client/fx.py:
```diff
@@ -68,6 +68,46 @@ def __init__(self, serial: str, capabilities: dict, daemon_dbus=None, matrix_dim
+
+        :return: Effect colors
+        :rtype: bytearray
+        """
```
Please add example values to the doc

tildearrow replied:
`grep -ri example` under pylib/ returns nothing. What is this supposed to be?

z3ntu replied:
Replied in https://github.com/openrazer/openrazer/pull/888#discussion_r365511694

tildearrow replied:
Done

z3ntu commented on pylib/openrazer/client/fx.py:
```diff
@@ -68,6 +68,46 @@ def __init__(self, serial: str, capabilities: dict, daemon_dbus=None, matrix_dim
+
+        :return: Effect speed
+        :rtype: int
+        """
```
Please add example values to the doc

tildearrow replied:
Done

z3ntu commented on pylib/openrazer/client/fx.py:
```diff
@@ -68,6 +68,46 @@ def __init__(self, serial: str, capabilities: dict, daemon_dbus=None, matrix_dim
+        :return: Wave direction
+        :rtype: int
+        """
+        return self._lighting_dbus.getWaveDir()
```
Please add example values to the doc

tildearrow replied:
Done

tildearrow committed:
= {} (without the newline) please

tildearrow committed:
Merge branch 'master' of https://github.com/openrazer/openrazer

tildearrow committed:
Merge branch 'master' of https://github.com/openrazer/openrazer

tildearrow committed:
Merge branch 'master' of https://github.com/openrazer/openrazer

tildearrow committed:
Merge branch 'master' of https://github.com/openrazer/openrazer

tildearrow committed:
Merge branch 'master' of https://github.com/openrazer/openrazer

tildearrow committed:
Merge branch 'master' of https://github.com/openrazer/openrazer

tildearrow committed:
*sigh* Luca I am getting tired of this

tildearrow committed:
probably finish DPI storage

TODO:
- fix pylint
- maybe fix hotplug (it is likely to fail)
- Luca-requested pylib changes
- make Luca smile for once :l

tildearrow committed:
......

tildearrow committed:
Merged #1001 into master.

tildearrow committed:
finally work around pylint and pass CI

z3ntu if you are about to request changes due to the workaround's ugliness:
- no! read the note in daemon/openrazer_daemon/hardware/device_base.py first.

tildearrow committed:
update get_(backlight|logo|scroll)_active funcs

to use local storage instead of accessing the hardware

tildearrow committed:
implement restoreLastEffect

TODO: test ripple

tildearrow committed:
update the pylib for restoreLastEffect

tildearrow committed:
add restoreLastEffect support to custom_starlight

tildearrow committed:
i'm done

tildearrow committed:
fix hotplug

tildearrow committed:
fix brightness return value

tildearrow committed:
compete with nightsky30

come on Luca

z3ntu committed:
Revert "finally work around pylint and pass CI"

This reverts commit ca5ab1d5862d1919d61ec79a593ac4d10ae3770a.

z3ntu committed:
Remove useless header

tildearrow committed:
Merge branch 'master' of https://github.com/openrazer/openrazer

tildearrow committed:
t push
:w
Merge branch 'master' of https://github.com/openrazer/openrazer

tildearrow committed:
fix getEffect for bw2013

tildearrow committed:
Merge branch 'master' of https://github.com/openrazer/openrazer

tildearrow committed:
Merged #978 into master.

tildearrow committed:
Merge branch 'master' of https://github.com/openrazer/openrazer

tildearrow committed:
Merge branch 'master' of https://github.com/openrazer/openrazer

lah7 committed:
Resolve conflicts

lah7 committed:
Separate device storage from razer.conf

- Keeps the same comments from example file
- More obvious where the persistence data is stored.

lah7 committed:
Remove unwanted new line
